### ISSUE TITLE
[css-grid-3] Should there be an `inline-grid-lanes` legacy keyword for `display`?

### ISSUE THREAD

**Issue by keithamus** [w3c/csswg-drafts#10961]
As per title; I haven't seen examples of `display: inline masonry`. I also think should ensure to include the legacy `inline-masonry` keyword if we're using `display`.

**Comment by Loirooriol** [comment]
We didn't add a legacy `inline-list-item` (dropped in #1495), `block-ruby` nor `block-math` (#5385), you need to provide 2 keywords for these combinations.

So I'm not convinced that we need `inline-masonry`, authors can just use `inline masonry` instead.

**Comment by bfgeek** [comment]
> So I'm not convinced that we need `inline-masonry`, authors can just use `inline masonry` instead.

We should add it as a convenience (there isn't any reason not to). E.g. most developers will use just `inline-grid` and not `inline grid`, and would be unnecessarily surprised if `inline-masonry` isn't supported. We shouldn't make this more difficult unnecessarily.

Ian

**Comment by keithamus** [comment]
`list-item` is special and `block` is assumed; the only legacy variants are `inline-` prefixed. While
I’m indifferent about adding `inline-masonry` I was largely pointing it out for completeness otherwise IIRC it would be the only inside display value without an inline legacy variant.

**Comment by Loirooriol** [comment]
We shouldn't call these "legacy" if we are going to keep adding them whenever we add new inside display types.

**Comment by Loirooriol** [comment]
> it would be the only inside display value without an inline legacy variant

Actually not, Houdini Layout adds [`layout()`](https://drafts.css-houdini.org/css-layout-api/#valdef-display-layout) but not `inline-layout()`.
It was removed by Ian in https://github.com/w3c/css-houdini-drafts/commit/60ff1c223a8eca286212fb9c6fb292344af7716a

**Comment by Loirooriol** [comment]
Renamed since now it's `grid-lanes` instead of `masonry`.

Agenda+ to decide whether to add a "legacy" `inline-grid-lanes` keyword.

**Comment by SebastianZ** [comment]
I've strongly in favor of _not_ introducing a new legacy keyword. That allows to push the awareness of the double-keyword syntax of `display`.

Sebastian

**Comment by jensimmons** [comment]
Making these work:
```
display: inline-flex;
display: inline-grid;
display: inline flex;
display: inline grid;
display: inline grid-lanes;
```
and making this broken:
```
display: inline-grid-lanes
```

is not a good way to help developers understand what's happening. 

I understand the desire to push developers into two-value syntax. The two-value syntax is better. But convincing developers to use it needs to come from teaching resources and news that it exists. 

Just making `display: inline-grid-lanes` invalid is only going to confuse people. It's going to make them think Grid Lanes doesn't work as an inline container. That it has to be block. 

> Authors over implementors over specifiers over theoretical purity.

**Comment by Loirooriol** [comment]
@jensimmons I fail to see how this argument applies to `inline-grid-lanes` but not to `inline-layout()`, `inline-list-item`, `inline-block-list-item`, `flow-root-list-item`, `block-math` and `block-ruby`, which we avoided in the past.

**Comment by Crissov** [comment]
Exactly, there are four legacy keywords in [CSS Display](https://drafts.csswg.org/css-display/#legacy-display) for `display`, all starting with ` inline-` and one of which is not simply “substitute hyphen by space”, but there are also other keywords that expand to multiple ones of which one is `inline`. I’d say that teaching/learning/understanding this is already complex and not at all helped by yet another alias. 

~~~~ css
display: inline;
display: inline flow;

display: inline-block;
display: inline flow-root;

display: inline-table;
display: inline table;

display: inline-flex;
display: inline flex;

display: inline-grid;
display: inline grid;

display:        ruby;
display: inline ruby;

display: inline      list-item;
display: inline flow list-item;

display: /* ./. */;
display: inline grid-lanes;
~~~~

**Comment by romainmenke** [comment]
To get authors familiar with the multi-keyword syntax it might help to have a Stylelint rule: https://github.com/stylelint/stylelint/issues/8958

**Comment by Crissov** [comment]
It might also help to make layout (i.e. formatting context) specifications talk about inside `display` values, especially in their headings and leave legacy shorthands (that are not shorter) to prose. 

| Spec | Now | Then |
| ---- | --- | ---- |
| [css-display](https://drafts.csswg.org/css-display-4/#inner-model) | 2.2. Inner Display Layout Models: the `flow`, `flow-root`, `table`, `flex`, `grid`, and `ruby` keywords | ✅ |
| [css-grid](https://drafts.csswg.org/css-grid/#grid-containers) | 5.1. Establishing Grid Containers: the `grid` and `inline-grid` `display` values | 5.1. Establishing Grid Containers: the `grid` inner `display` value |
| [css-flexbox](https://drafts.csswg.org/css-flexbox/#flex-containers) | 3. Flex Containers: the `flex` and `inline-flex` `display` values | 3. Establishing Flex Containers: the `flex` inner `display` value |
| [css-table](https://drafts.csswg.org/css-tables-3/#table-structure) | 2.1. Table Structure | *2.1.1. Establishing Table Containers: the `table` inner `display` value(?) |
| [css-ruby](https://drafts.csswg.org/css-ruby/#formatting-context) | 2.1.1. The Ruby Formatting Context | 2.1.1. The Ruby Formatting Context: the `ruby` inner `display` value |
| [MathML](https://w3c.github.io/mathml-core/#new-display-math-value) | 4.1 The `display: block math` and `display: inline math` value | 4.1 Establishing Math Containers: The `math` inner `display` value |
| [css-houdini](https://drafts.css-houdini.org/css-layout-api/#layout-api-containers) | 2. Layout API Containers | *2.1. The `layout()` function for inner `display` layout models (?) |

**Comment by bfgeek** [comment]
So I think a lot of the people that are in support of adding legacy variants of display keywords weren't on the call when this was discussed unfortunately.

Broadly speaking the majority developers don't know that the syntax `display: inline flex` exists.

A search on github:
`display: inline grid` : https://github.com/search?q=%22display%3A+inline+grid%22&type=code ~1k results
`display: inline-grid` : https://github.com/search?q=%22display%3A+inline-grid%22&type=code ~134k results

`display: inline flex` : https://github.com/search?q=%22display%3A+inline+flex%22&type=code ~2.5k results
`display: inline-flex` : https://github.com/search?q=%22display%3A+inline-flex%22&type=code ~2.7M results

I suspect that a HTTP archive search will reveal similar results, broadly speaking there is a 2-3 *orders of magnitude* difference in usage. Put differently a developer has a 0.1-0.01% change that they'll see this, and due to how the human mind reads, they may not even notice the difference while reading source code.

As another data-point lots of developers use LLMs for answering questions, I tested 2 of the top LLM engines with the query:
"In CSS, how do I make a flexbox sit on the same line as some text?"
And both responded to use `display: inline-flex`, with no mention of the `inline` keyword variant.
 
It is entirely reasonable for a developer who has used `display: inline-flex` to place a flexbox on a while to expect `display: inline-grid-lanes` to work. They've done this their whole career so far! And its just worked!

Breaking this convenience for spec purity is a large disservice to web developers. It costs browser engines nothing[1] to add this feature, and doesn't harm CSS as a language going forward.

Ian

[1] Relatively speaking

**Comment by SebastianZ** [comment]
@bfgeek:

> Breaking this convenience for spec purity is a large disservice to web developers. It costs browser engines nothing[1] to add this feature, and doesn't harm CSS as a language going forward.

This is not at all about spec purity. It is about educating authors that there is a two-value syntax.
And the two-value syntax is still quite new, while the single-value keywords are around for much longer. So, of course, the adoption is still much lower. And that low adoption rate is also the reason why LLMs still suggest the single-keyword values.

@jensimmons:

> Just making `display: inline-grid-lanes` invalid is only going to confuse people. It's going to make them think Grid Lanes doesn't work as an inline container. That it has to be block.

I'd say it is more likely that they look it up and check why `inline-grid-lanes` is not working and find those articles that explain the two-value syntax. Or they ask their favorite LLM that tells them so.

And teaching resources and articles aside, an auto-completion of multiple keywords in tools would probably already go a long way.
Furthermore, once we turn `display` into a shorthand for inner and outer display values (see #4729 and related ones), the two-value syntax will surely see a huge increase in adoption.

Single-value vs. two-value syntax aside, I agree with @kizu who noted that there probably aren't that many use cases for an inline grid-lanes layout, anyway. And as he noted, "if you actually need inline grid-lanes, you probably know a lot about how display works".

Sebastian

**Comment by bfgeek** [comment]
> This is not at all about spec purity. It is about educating authors that there is a two-value syntax.

Educating authors should be done by educating, not by withholding a useful feature. Not providing this only causes difficulty to well meaning web developers.

**Comment by kizu** [comment]
One note about the current usage: I am not sure if it shows anything, as even people who know about the multiple keywords syntax won't necessarily use it in practice. Why use something that works the same but has [much worse](https://caniuse.com/mdn-css_properties_display_multi-keyword_values) browser support?

If anything, in my practice, I remember wanting to use the longhands for inner and outer displays separately and not being able to, but I don't remember needing to use the shorthand with two values.

- - -

Another similar comparison of a newer syntax vs. a legacy syntax and not maintaining the legacy syntax for the new thing: colors.

Usage-wise, it is a similar case to `rgba()` with spaces & slash syntax compared to the commas:

- `"rgba(0, 0"`: https://github.com/search?q=language%3ACSS+%22rgba%280%2C+0%22&type=code — 15M files
- `"rgba(0 0"`: https://github.com/search?q=language%3ACSS+%22rgba%280+0%22&type=code — 11K files.

And it is also a case, where I don't see any reason to use the new syntax in practice, as it has clearly worse browser support.

But for the new color functions like `oklch()` we introduced _only_ the new syntax, without the support of the legacy one. And people did learn it and started using it.

It will be consistent to not introduce `inline-grid-lanes` if we consider these “short” values legacy.

The Web Platform Design Principles say, [“Leave the web better than you found it”](https://www.w3.org/TR/design-principles/#leave-the-web-better).

> Consistency with existing problematic design patterns is not a good reason to propagate those patterns in newly specified Web APIs or platform features. Where possible, build new web capabilities that improve the overall platform quality by mitigating existing defects.

Do we consider having a “short” value (that is not shorter) alongside two keywords a problematic design pattern? If we were to add another 10 different display types, would we continue adding duplicate values for consistency?

Another principle says, [“Prefer simple solutions”](https://www.w3.org/TR/design-principles/#simplicity).

> Simple solutions are generally better than complex solutions, although they may be harder to find. Simpler features are easier for user agents to implement and test, more likely to be interoperable, and easier for authors to understand. It is especially important to design your feature so that the most common use cases are easy to accomplish.

What is simpler: having two alternatives for the same thing, or just one? Is this added complexity granted by its use case, or do we have the most common use cases covered by regular `grid-lanes`?

**Comment by romainmenke** [comment]
I was very surprised by this resolution and the arguments that led to it.

- historical data will always favor legacy syntax, using that as an argument to keep doing the legacy thing will block any progress
- llm's just regurgitate whatever they have seen the most
- any actual feedback from authors was seemingly ignored, while hypotheticals where given weight

--------

Obviously historical data will indicate that `inline-flex` is much more widely used than `inline flex`. There never was an incentive to use `inline flex` but there always was a reason to use `inline-flex`, for better browser version support. This sort of stuff trickles down into articles, guides, stackoverflow posts, llm training data, ... and takes a very long time to change.

This doesn't imply that authors are opposed to `inline flex` or that they would have issues with only `inline grid-lanes`.

Authors might collectively be choosing to use a worse syntax (`inline-flex`) for the good of end users who are still on older browser versions.

Instead of speculating about this, isn't it better to gather actual data?

Multi-keyword syntax has only recently become Baseline Widely Available. Devrel efforts to promote its usage can only now really start. I'd prefer it if we gave this a little more time before drawing conclusions. Authors might surprise us.

--------

If you ask ChatGPT (hadn't touched that monstrosity in years...) to create a masonry layout it will also spew out that which it has been trained on, which are some of the articles around earlier designs:

```
.masonry {
  display: grid;
  grid-template-columns: repeat(auto-fill, minmax(250px, 1fr));
  grid-template-rows: masonry;
  gap: 1rem;
}
```

If you then point it to the latest draft spec it will use `display: grid-lanes;`.
If you ask it to make it `inline` it will use `inline-grid-lanes` because that is explicitly mentioned in the spec.
If you ask it to use the multi keyword syntax to make it `inline` it will use `inline grid-lanes`.

So it seems that llm's will just produce for authors whatever we spec, write about, document,...

If we remove `inline-grid-lanes` from the spec I have no doubt that an llm will suggest `inline grid-lanes`, as it seems capable of this.

--------

What I understand least of all is that some of us are concerned that authors will be upset about this but there doesn't seem any source for this concern?

Has there been a poll or article about this or some way to gather this feedback?
From the minutes this seems to have been an impression, but not something that was actually tested.

### poll by kizu

While @kizu [did a poll](https://front-end.social/@kizu/115947014705147754) that seemed to favor not adding `inline-grid-lanes`. It might not have had massive reach and I can imagine that people who follow @kizu are already more knowledgeable about CSS than others, it is still a real data point.

### MDN docs

MDN warns against using multi keyword syntax:

https://developer.mozilla.org/en-US/docs/Web/CSS/Reference/Properties/display

> Note: To be sure layouts work on older browsers, you may use single-value syntax, for example display: inline flex could have the following fallback

https://developer.mozilla.org/en-US/docs/Web/CSS/Guides/Display/Multi-keyword_syntax#what_about_the_old_values_of_display

> The single values of display are described in the specification as legacy values, and currently you gain no benefit from using the multi-keyword versions, as there is a direct mapping for each multi-keyword version to a legacy version, as demonstrated in the table above.

So that isn't great if we want authors to adopt the multi-keyword syntax...
This goes back to the incentives for authors.

Given that the multi-keyword syntax is now Baseline Widely Available we could update those docs to caution less against it.

### Stylelint rule

Over at Stylelint we've been rushing to get a rule out, given this thread.

Our intention was to create awareness of the multi-keyword syntax and promote its usage.
Thereby making it easier to introduce only `inline grid-lanes`.

This seemed a good time for this because:
- the syntax has been supported in all browsers since mid 2023
- popular minifiers and transpilers will convert `inline flow-root` into `inline-block` anyway

What I didn't expect was the early feedback we got there.

It seems that at least one author really wants their colleagues to only use the full multi-keyword syntax. See: https://github.com/stylelint/stylelint/issues/8958#issuecomment-3759008709

They reason that it is much easier to understand how the `display` property works if you use the full syntax.

I am less in favor of using the full notation over shorter values but I do agree with the overal point they're making.

It is something I have also observed, that colleagues will use `inline-flex` when any `inline` would have worked, just because that is what worked last time. By making it one keyword authors seem to associate traits with a specific precomposed value instead of with either an inner or outer display mode.

We hoped to ship this rule and make it a default in a commonly used preset. This preset would either autofix to the full notation in all cases or only for the legacy/precomposed values. When confronted with this, authors would definitely tell us if and how angry they are about it.

Not every CSS author uses Stylelint, but such signals are still better than pure speculation.

--------


The few actual author signals seem to favor the previous resolution.
I don't understand why we reversed it without real data.

I am in favor of going back to the previous resolution and gathering author feedback first.

There is also no rush here. The first browsers to ship this could only ship `inline grid-lanes`. If this makes authors angry we can still add `inline-grid-lanes` in time before there is interop.

**Comment by Crissov** [comment]
> it doesn't cost anything for browsers to add this

If that argument had ever worked, CSS would have quite a number of convenience features for authors.

**Comment by astearns** [comment]
> ### Stylelint rule
> 
> Over at Stylelint we've been rushing to get a rule out, given this thread.
> 
> Our intention was to create awareness of the multi-keyword syntax and promote its usage. Thereby making it easier to introduce only `inline grid-lanes`.

FWIW I think you could still introduce this rule, and it would still be a good idea.

**Comment by benface** [comment]
> Furthermore, once we turn `display` into a shorthand for inner and outer display values (see [#4729](https://github.com/w3c/csswg-drafts/issues/4729) and related ones), the two-value syntax will surely see a huge increase in adoption.

I truly believe this to be true. How to get the ball rolling on #4729? I'd like to mention #12046 as well, as I think there is value in being able to toggle `display: none` independently from the "outside display".

**Comment by SebastianZ** [comment]
While there was no one in the meeting to object the resolution, I do speak out a formal objection to the [last resolution](https://github.com/w3c/csswg-drafts/issues/10961#issuecomment-3820508516) now.

@romainmenke @kizu @astearns @emilio @Loirooriol and others being against introducing `inline-grid-lanes` all brought up good arguments plus tried to provide some data, while the people _for_ introducing a single keyword only restated the arguments that were either already refuted or purely hypothetical.

I strongly believe it is counter-productive to introduce the `inline-grid-lanes` and remove the indication that the single `inline-*` keywords are legacy.

The main argument for introducing a single keyword was to avoid tripping up authors. Though that is only a hypothesis. And even if that were true, there will be plenty of sources teaching them that they need to use space separation in that case.

> There is also no rush here. The first browsers to ship this could only ship `inline grid-lanes`. If this makes authors angry we can still add `inline-grid-lanes` in time before there is interop.

Note that my objection only refers to _hastily_ adding the legacy variant and declaring the `inline-*` keywords as _not_ being legacy.
As Romain wrote, when there's a strong complain from authors about a missing `inline-grid-lanes` and people are not happy when tought that they should use the two-value syntax instead, we can still introduce it at a later point.
Though I'd claim that _not_ introducing the single-keyword variant pushes the awareness of inner and outer display types and the adoption of the two value syntax, which is more benefitial in the long term.

> it doesn't cost anything for browsers to add this

Requests for introducing aliases are normally declined by the WG arguing that those are very expensive to maintain, as they not only mean implementation adjustments but also spec, testing, and documentation changes.
That's also why we don't have a `no-wrap` alias or `current-color`, `corner-radius`, or multi-keyword color names using dashes.

But if the group changed their mind about that, I'd be in favor of introducing those aliases.

Sebastian

**Comment by astearns** [comment]
Removed the Needs Edits tag based on the objection from @SebastianZ.

While I’m not a fan of the most recent resolution here and would prefer to wait on introducing `inline-grid-lanes`, I don’t think the presence or absence of it will do much for or against use of the two-value display syntax.

**Comment by romainmenke** [comment]
The Needs Edits tag was added for the previous resolution: https://github.com/w3c/csswg-drafts/issues/10961#issuecomment-3720021476

So I guess those still need to be applied?

**Comment by Loirooriol** [comment]
I'm not sure what's the status now, some people were strongly wanting the keyword, do we need to confirm if someone is objecting to removing the keyword from the specs? Or just start removing it as per Sebastians's objection?

Apple has already enabled it in Safari Technology Preview, not sure how much time until it ships.

**Comment by astearns** [comment]
We took the resolution to include `inline-grid-lanes` because proponents were strongly in favor and no one objected. Now we have an objection to that resolution.

@bfgeek @jensimmons (or any other folks in favor of adding the value) would you object to going back to the previous resolution (wait until we have compelling evidence before adding the value)?

**Comment by SebastianZ** [comment]
> The Needs Edits tag was added for the previous resolution: [#10961 (comment)](https://github.com/w3c/csswg-drafts/issues/10961#issuecomment-3720021476)

Independent from my objection, edits are needed either way. Either we remove `inline-grid-lanes` from the spec (and WebKit's implementation) as per the first resolution or we remove the notes related to the single-keyword values being legacy as per the second resolution. So adding the label back.

And just to emphasize, I don't want to block shipping this feature. I just don't believe, introducing new single-keyword aliases is the better path forward.

Sebastian

**Comment by romainmenke** [comment]
Small update on improving the authoring experience and docs for the multi-keyword syntax:
- [the Stylelint rule](https://stylelint.io/user-guide/rules/display-notation) has shipped, but we haven't added it to a preset (yet)
- [mdn display docs](https://developer.mozilla.org/en-US/docs/Web/CSS/Reference/Properties/display) have been updated so that they don't warn against using the multi-keyword syntax. also added an example of the multi-keyword syntax in the first demo
- vscode syntax highlighting was missing for the `flow` keyword, this has been added and will ship in an upcoming release

**Comment by Loirooriol** [comment]
Tracking issues for removing `inline-grid-lanes`:

 - Blink: https://issues.chromium.org/issues/474084992
 - WebKit: https://bugs.webkit.org/show_bug.cgi?id=307255

Gecko hasn't landed the patch introducing the keyword (https://bugzilla.mozilla.org/show_bug.cgi?id=1995348)

**Comment by bfgeek** [comment]
> would you object to going back to the previous resolution (wait until we have compelling evidence before adding the value)?

@astearns - I think this really needs to be discussed on a call.

@Loirooriol - There is a lot more to a formal objection than just "we go back to the previous resolution". Filing those issues is a little pre-emptive. I've closed Blink's until we get clarity here (at which point our CSSWG bot will file a new issue against Blink).

Ian

**Comment by astearns** [comment]
FWIW, I’m interpreting @SebastianZ's comment as a regular objection for the moment. I think it’s premature to invoke the formal objection process.

@bfgeek one of the things that will come up on a call is whether there are sustained objections to the previous resolution, so I’m trying to save time by discussing this here first.

**Comment by bfgeek** [comment]
> one of the things that will come up on a call is whether there are sustained objections to the previous resolution, so I’m trying to save time by discussing this here first.

I personally don't like to throw around objections (at least I try not to), but my read of the room in the previous meeting that they the people advocating for a `inline-` prefix had relatively strong opinions on it.

**Comment by bfgeek** [comment]
> I think it’s premature to invoke the formal objection process.

Agreed - but words have meanings :).

**Comment by astearns** [comment]
My read is that we have strong opinions on both sides. So I’m looking to see what new information we can apply to make a good (or at least non-bad) decision here.

We previously resolved that we should wait until we have compelling evidence that we should introduce this value. That resolution passed without objection.

Then we reversed that decision, but now have an objection to that second resolution.

Lacking objections to the first resolution, I’m inclined to let it stand. That’s the reason I’ve asked.

**Comment by tabatkins** [comment]
I think we have implementors wanting to implement the value based on observations of user patterns (for the values that appear in both forms, both users and guides overwhelmingly still favor the single-keyword version), and objections coming from non-implementors based on theoretical purity (it's less cluttered and reinforces the inside/outside concept separation).  

Given that the *actual cost* of it is effectively nil and happily borne by the implementors in question, I think this falls pretty squarely under the `user > author > impl > theory` rubric, and the WG should override the objections. Chrome wishes to maintain the resolution we made at the meeting. (And WebKit has already shipped the value, so presumably they feel the same.)

**Comment by tabatkins** [comment]
(And yes, I recognize that "teachability" or similar concerns awkwardly span across the divide, touching both "theory" *and* "user". But in the absence of demonstrated or convincingly argued user *harm*, and in the presence of widespread uptake, I think objections along that axis are much more strongly on the "theory" side.)

**Comment by Loirooriol** [comment]
I don't share the opinion that this makes it better for authors. As mentioned beforehand, authors have had no problem at all with the new commaless color syntax, so I don't see why they would have a problem here. And thus I disagree with overriding the objection on the basis of `user > author > impl > theory`.

**Comment by romainmenke** [comment]
> both users and guides overwhelmingly still favor the single-keyword version

Up until very recently the multi-keyword syntax wasn't Baseline Widely Available and the docs actively warned against using the multi-keyword syntax as a result of that.

I don't think we can make the assumption that ~~users~~ authors favor the `inline-*` syntax. I think they didn't consider the multi-keyword syntax ready for production use. It takes a really long time before authors voluntarily switch to a syntax that has worse browser support, especially if it doesn't offer new capabilities. This however doesn't necessarily mean that they don't want to use the more modern syntax or are unaware of it.

I think we need to recognize that we can't get a clear signal from authors about what they favor at this point in time, purely by looking at past behavior.

I am not saying that authors don't want `inline-grid-lanes`. I am sure that some authors will try the `-` and be surprised if it doesn't work.

Personally I don't feel very strongly about adding or not adding `inline-grid-lanes`.
Anecdotal evidence (from authors!) seems to indicate that it is fine to not add this but other than that I am fine with adding it.

------

Can we do a compromise?
- add `inline-grid-lanes`
- keep all the `inline-*` (including `inline-grid-lanes`) marked as legacy syntax
- decide not to add `inline-*` variants in the future, with the option to revisit later

When there is a new display mode in ±5 years we will have a much clearer view on what authors favor and how aware they are of the multi-keyword syntax.

Marking these as legacy helps to promote the multi-keyword syntax in documentation, lint rules, ... If these are not marked as legacy it is much harder to drive adoption.

Authors can learn about `inline grid-lanes` and thereby also discover the multi-keyword syntax. But those that just try `inline-grid-lanes` won't get surprises.

(Keeping in mind that we are not considering a deprecation or removal of `inline-*`, only an indicator that this syntax is legacy.)

**Comment by benface** [comment]
> When there is a new display mode in ±5 years we will have a much clearer view on what authors favor and how aware they are of the multi-keyword syntax.

The problem is, there is no reason to favor the multi-keyword syntax. It's not shorter, and it doesn't provide any benefit as far as I know. Not having `inline-grid-lanes` would be a good reason to favor it: consistency in a codebase, and even if one doesn't use grid lanes, future-proofing against inconsistencies (since every possibility can be expressed with multi-keyword but not with single-keyword).

Anyway, looks like the ship might have sailed.

**Comment by kizu** [comment]
A note: I think (please, let me know if I am mistaken) some of the above posts conflate user & author.

For the _user_ there is no difference if it is `inline grid-lanes`, `inline-grid-lanes` or both. It works the same.

For the _author_, yes, there is a difference, and I see arguments in favor of both ways to resolve this. However, I think the suggestion that “authors will be angry” with the multi-keyword syntax is as theoretical as any kind of theoretical purity the multi-keyword syntax favors.

I don't consider there to be a violation of “user > author > impl > theory” because we only speculate about the authors’ position. Unless we consider the invited experts’ positions to be closer to it. I am no longer one (although, tbh, nothing really changed after I became a regular member), but @SebastianZ is one.

If we care about what authors actually say — we need to query and listen to them. If we think this is such a small issue that we don't want to waste time on this — then we can't put words into their mouths and speculate.

**Comment by kbabbitt** [comment]
> Can we do a compromise?
> 
> * add `inline-grid-lanes`
> * keep all the `inline-*` (including `inline-grid-lanes`) marked as legacy syntax
> * decide not to add `inline-*` variants in the future, with the option to revisit later

I was thinking along the same lines myself and would support such a compromise. Leaning on the TAG design principle of [consistency](https://www.w3.org/TR/design-principles/#consistency):

> Since the web platform has gradually evolved over time, there are often multiple conflicting precedents which are mutually exclusive. You can weigh which precedent to follow by taking into account prevalence (all else being equal, follow the more popular precedent), API ergonomics (all else being equal, follow the more usable precedent), and API age (all else being equal, follow the newer precedent).

As a general principle, I think it makes sense to lean into the multi-keyword syntax, because it's more flexible. Variable substitution can be used for one term or the other, for example. That makes it more ergonomic. Going way back in the conversation, I agree we don't need combined values for any of the following:
> We didn't add a legacy inline-list-item (dropped in https://github.com/w3c/csswg-drafts/issues/1495), block-ruby nor block-math (https://github.com/w3c/csswg-drafts/issues/5385), you need to provide 2 keywords for these combinations.

However, it is also the case that `inline-grid` is way more prevalent then `inline grid` in current usage. I think it's reasonable for authors to expect that if they go in and append `-lanes`, it will work. So I would be in favor of introducing `inline-grid-lanes` for the specific reason that grid-lanes shares a name with a layout that already has a combined value.

**Comment by astearns** [comment]
Looking through the minutes, it seems that @Loirooriol and @fantasai were arguing most strongly for removing the legacy syntax classification. To both of you, would you be OK with @romainmenke's compromise of leaving the new `inline-grid-lanes` as legacy for now and postponing the legacy versus two-value syntax question until the next time it comes up?

**Comment by tabatkins** [comment]
Yes, we can (and often should) break precedent in favor of something better, when the old style was bad for whatever reason. That's not the case here. The difference is literally whether we spell the value with a space or a dash, and the proposal is that we continue to allow both, following existing precedent. The `inline-grid` precedent, *specifically*, is a strong consistency argument for having `inline-grid-lanes` as well.

And it is of course important to remember that the argument is not "do we have dashed keywords, or spaced keywords". It's just "do we (continue to) have dashed keywords" - the space-separated syntax is already there and stable, with nobody suggesting any action around it. Whatever arguments one might have for why the space-separated syntax is *superior* are irrelevant for this argument, unless one is trying to make the (imo much more difficult to support) argument that having both forms is *confusing* or otherwise *harmful* to authors in some way.

**Comment by tabatkins** [comment]
Yeah I don't care particularly how we classify the inline-* keywords. I do agree with Oriol and Fantasai that it's a little silly to call them legacy if we're still adding them, but it doesn't bother me to keep them as they are.

**Comment by Loirooriol** [comment]
TBH I'm still not really understanding what's special about `inline-grid-lanes` vs the other keywords that we didn't add. If the difference is popularity, then will people also want to keep adding "legacy" keywords for all new popular inner display types?

If this is just an exception for consistency with `inline-grid`, ok.

**Comment by SebastianZ** [comment]
> If this is just an exception for consistency with inline-grid, ok.

The consistency with `inline-grid` is the only argument that seems reasonable for me that authors might expect `inline-grid-lanes` to work.

I still believe that adding `inline-grid-lanes` is unnecessary and harmful for pushing the two-keyword syntax, though @romainmenke's [proposed compromise](https://github.com/w3c/csswg-drafts/issues/10961#issuecomment-3879948345) is acceptable for me. So I lift my objection if we can agree on that.

Sebastian

**Comment by Loirooriol** [comment]
> unless one is trying to make the (imo much more difficult to support) argument that having both forms is confusing or otherwise harmful to authors in some way

I'm not saying that much, but note that serialization prefers the legacy one, so adding both relegates the new (better) syntax as some kind of 2nd class citizen. This was fine when required by compat, but that's not the case here.

**Comment by Crissov** [comment]
So if it had been `masonry` or `semi-grid` instead of `grid-lanes`, people would be fine without an `inline-`-prefixed legacy-ish keyword?! 🤨

**Comment by astearns** [comment]
The CSSWG will automatically accept this resolution one week from now if no objections are raised here. Anyone can add an emoji to this comment to express support. If you do not support this resolution, please add a new comment.

Proposed Resolution:  Keep all inline-* values (including inline-grid-lanes) marked as legacy syntax

**Comment by mirisuzanne** [comment]
I do think the two value syntax is better both for author understanding and for forward flexibility. I agree with @SebastianZ that our decision here was unnecessary and harmful for moving that direction, and with @Loirooriol that it's strange to favor the legacy spelling of a new mode that doesn't need that spelling. Is there any way we can take the compromise further, and favor the more modern spelling of new displays?

**Comment by kbabbitt** [comment]
> I do think the two value syntax is better both for author understanding and for forward flexibility. I agree with [@SebastianZ](https://github.com/SebastianZ) that our decision here was unnecessary and harmful for moving that direction, and with [@Loirooriol](https://github.com/Loirooriol) that it's strange to favor the legacy spelling of a new mode that doesn't need that spelling. Is there any way we can take the compromise further, and favor the more modern spelling of new displays?

What if we add something like:

> [!Note]
> In general, the CSSWG intends to support new display modes with the two value syntax only,
> as it is more flexible than the legacy precomposed display values.
> `inine-grid-lanes` was added as an exceptional case due to the prevalence of `inline-grid`.

**Comment by mirisuzanne** [comment]
I like the note. Any chance we can avoid using the 'legacy' option for serialization?

**Comment by fantasai** [comment]
> it's a little silly to call them legacy if we're still adding them

This.

> In general, the CSSWG intends to support new display modes with the two value syntax only,

This was our intention ten years ago, and ten years is a long time ago. If we're not willing to go along with that intention now, why does it make sense to keep that intention on the books? What makes us think that we would follow through on that intention in the future, if we aren't going to follow through on it now?

> I'm not saying that much, but note that serialization prefers the legacy one, so adding both relegates the new (better) syntax as some kind of 2nd class citizen. 

Just wanted to highlight this point by @Loirooriol. It doesn't make sense to say in prose that the two-keyword syntax is preferable and the one-value syntax is discouraged while also continuing to add new hyphenated versions and making the OM auto-convert the two-keyword values to (supposedly legacy) hyphenated ones.

**Comment by astearns** [comment]
I don’t know whether we’ll follow through on the intent for future display values. But I don’t think it’s harmful to defer the question until it comes up next. I’m looking for compromise between people convinced we need this particular `inline-*` value and those unhappy that we have not followed through with our longstanding intent.

One reason for this current exception is the correspondence with `inline-grid` - perhaps in the future that won’t be an issue.

Another reason for this current exception is that the available data hasn’t shown author uptake of the two-value form, in part because it took a while for the two-value form to be widely available. Perhaps in the future the author usage data will shift.

I think there are enough folks that don’t want to give up on the two-value intent just yet, that taking this one value as an exception is a reasonable compromise.

How `inline-grid-lanes` serializes could perhaps be a new, separate issue.

**Comment by fantasai** [comment]
I don't see how any of the arguments made to support `inline-grid-lanes` here wouldn't be equally valid next time this comes up.

**Comment by romainmenke** [comment]
Then we also have to revisit `inline-list-item`, `block-math`, `block-ruby` as the same arguments would also apply to those keywords?

**Comment by astearns** [comment]
@fantasai you disagree with the proposed async resolution. For clarity, do you object to us taking this resolution?

If so, I think the next step would be to find some other way to get past @SebastianZ’s objection to the previous resolution.

**Comment by SebastianZ** [comment]
> @fantasai you disagree with the proposed async resolution. For clarity, do you object to us taking this resolution?

My interpretation of @fantasai's last comments is that she's advocating for consistency in regard of favoring the two-value syntax, while they are ambigious regarding whether we should continue adding single-keyword aliases or not.
Unfortunately, I can't see a way to fulfill that criteria together with the proposed resolution.

> > it's a little silly to call them legacy if we're still adding them
> 
> This.

That's true. And it is one reason people didn't like the alias to get added in the first place.

> > In general, the CSSWG intends to support new display modes with the two value syntax only,
> 
> This was our intention ten years ago, and ten years is a long time ago. If we're not willing to go along with that intention now, why does it make sense to keep that intention on the books? What makes us think that we would follow through on that intention in the future, if we aren't going to follow through on it now?

So this basically boils down to whether we want to continue advocating for the two-value syntax or not.
Currently, the group obviously disagrees in that point. And with the proposed resolution we tried to find a middleground.

I just want to restate the initial resolution we took:

> RESOLVED: remove inline-grid-lanes for now, but we are open to adding it again if there is evidence it solves a problem

This was already a good compromise, in my opinion, which would keep the consistency @fantasai is asking for. And I still didn't read a strong argument against that.
This resolution could be interpreted as "let's try to continue pushing the two-value syntax but if we're not successful, finally give up on that".
And I believe, we're not at the point to give up yet, especially as the two-value syntax is still pretty new to authors and we therefore simply don't know yet how the uptake of it will be in the near future.

So again, what's the reason to _not_ go with only the two-value syntax for now and see if people complain, and _then_ reconsider introducing new single-keyword values and with that give up on calling them legacy?

Sebastian

**Comment by fantasai** [comment]
I'm pretty strongly against us being knowingly dishonest in our specifications. And I believe we need to sort out the serialization question here, as it's a pretty integral part of this debate and not really a follow-up issue.

**Comment by astearns** [comment]
I think “knowingly dishonest” is vastly overstating the case.

**Comment by fantasai** [comment]
> Then we also have to revisit inline-list-item, block-math, block-ruby as the same arguments would also apply to those keywords?

@romainmenke No, I think we can reasonably say that the `inline-*` pattern is for inner display types only, and there's no `block-*` pattern.

**Comment by Loirooriol** [comment]
Even with just `inline-*`, it's not clear if the arguments supporting `inline-grid-lanes` also apply to `inline-list-item`, `inline-block-list-item` and `inline-layout()`.

**Comment by fantasai** [comment]
@Loirooriol So far it's been "all default-block-level inner display type values get `inline-*` keywords". If we add `inline-grid-lanes` that keeps the pattern. Not adding `inline-grid-lanes` creates a fixed, historically-determined set ("legacy"). Neither of these options requires adding all the extra values you're mentioning, that we don't have.

**Comment by Loirooriol** [comment]
`layout()` is an inner display type that defaults to block-level, so the pattern isn't working.

**Comment by SebastianZ** [comment]
With one engine already shipping this feature and at least one actively being worked on, I'd like to revive this discussion in the hope that we can find a viable solution before it's too late.

Repeating my previous statement:

> > RESOLVED: remove inline-grid-lanes for now, but we are open to adding it again if there is evidence it solves a problem
> 
> This was already a good compromise, in my opinion, which would keep the consistency [@fantasai](https://github.com/fantasai) is asking for. And I still didn't read a strong argument against that.

Reading through the comments again, it seems nobody actually gave an proper argument against this first resolution. (Please let me know if I missed it! This thread got quite long...)
As @kizu [stated earlier](https://github.com/w3c/csswg-drafts/issues/10961#issuecomment-3880346059), any negative reaction of authors for _not_ introducing a single-keyword value is speculative. And that resolution let the door open for introducing it _if_ there's request for it.

@fantasai wrote:

> [@Loirooriol](https://github.com/Loirooriol) So far it's been "all default-block-level inner display type values get `inline-*` keywords". If we add `inline-grid-lanes` that keeps the pattern.

This hasn't been written down anywhere before, has it? At least I couldn't find any resource using this definition as a reason for having the `inline-*` keywords.

> Not adding `inline-grid-lanes` creates a fixed, historically-determined set ("legacy").

The spec and different resources explaining the two-value syntax like [MDN](https://developer.mozilla.org/en-US/docs/Web/CSS/Reference/Properties/display#which_syntax_should_you_use) _already_ outline the `inline-*` keywords as "legacy". The [resolution I objected to](https://github.com/w3c/csswg-drafts/issues/10961#issuecomment-3820508516) asking to "remove [the] indication that inline- syntaxes are legacy or otherwise inferior" not only targets the spec but also those resources.

> I'm pretty strongly against us being knowingly dishonest in our specifications.

What are you actually referring to here?

> And I believe we need to sort out the serialization question here, as it's a pretty integral part of this debate and not really a follow-up issue.

@mirisuzanne wrote:

> Any chance we can avoid using the 'legacy' option for serialization?

I don't think serialization is such a big deal. Though if we want to keep advocating for the two-value syntax, I think it should serialize as two values even if we introduce the single-keyword variant.
In any case, it should be consistent for `inline grid-lanes` and `inline-grid-lanes`.

-----

So, let's bring this to the agenda again with the main questions being:

* Are there objections against the [first resolution](https://github.com/w3c/csswg-drafts/issues/10961#issuecomment-3720021476)?
* Do we want to keep advocating for the concept behind inner and outer values?
* Depending on the above, how to serialize the value?

Sebastian

**Comment by mirisuzanne** [comment]
I've been teaching the multi-value syntax for years, and authors find it useful for understanding the various things `display` is doing. But it won't see broad teaching & adoption until there is functionality that relies on it. Authors don't get a lot of training in CSS, they mostly read how-to articles about things they are trying to achieve. If we're not willing to put any features entirely in the new syntax, it will remain a useful curiosity that only some developers encounter.

I think the most attractive way to put functionality behind the multi-value syntax is with longhands — especially the much-needed `display-visibility` (name TBD) https://github.com/w3c/csswg-drafts/issues/12046 mentioned by @benface earlier.

I also think authors are entirely capable of learning that `inline grid-lanes` is the proper syntax, if we stick to the original resolution. And when they ask why, they will discover the multi-value syntax – and that it is broadly available. That's good for everyone.

To me, continuing with two syntaxes only increases an _inconsistency_ of the language. We're dragging out the point of confusion, rather than making a clean break. Deprecating the old syntax makes things more consistent and learnable than trying to keep both syntaxes alive _just in case someone tries the old one_ for a new feature.

(but I did and do not feel strongly enough to object)
